scylla-cdc/src/log_reader_builder.rs

```diff
@@ -0,0 +1,310 @@
+    end_timestamp_sender: tokio::sync::watch::Sender<Option<chrono::Duration>>,
+}
+
+impl CDCLogReaderBuilder {
+    pub fn new() -> CDCLogReaderBuilder {
+        let end_timestamp = None;
+        let (end_timestamp_sender, end_timestamp_receiver) =
+            tokio::sync::watch::channel(end_timestamp);
+        let session = None;
+        let keyspace = None;
+        let table_name = None;
+        let start_timestamp = chrono::Duration::from_std(
+            SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap(),
+        )
+        .unwrap();
+        let end_timestamp = None;
+        let window_size = Duration::from_millis(WINDOW_SIZE as u64);
+        let safety_interval = Duration::from_millis(SAFETY_INTERVAL as u64);
+        let sleep_interval = Duration::from_millis(SLEEP_INTERVAL as u64);
+        let consumer_factory = None;
+        let readers = vec![];
+
+        CDCLogReaderBuilder {
+            session,
+            keyspace,
+            table_name,
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            readers,
+            end_timestamp_receiver,
+            consumer_factory,
+            end_timestamp_sender,
+        }
+    }
+
+    pub fn session(mut self, session: Arc<Session>) -> Self {
+        self.session = Some(session);
+        self
+    }
+
+    pub fn keyspace(mut self, keyspace: &str) -> Self {
+        self.keyspace = Some(keyspace.to_string());
+        self
+    }
+
+    pub fn table_name(mut self, table_name: &str) -> Self {
+        self.table_name = Some(table_name.to_string());
+        self
+    }
+
+    pub fn start_timestamp(mut self, start_timestamp: chrono::Duration) -> Self {
+        self.start_timestamp = start_timestamp;
+        self
+    }
+
+    pub fn end_timestamp(mut self, end_timestamp: chrono::Duration) -> Self {
+        self.end_timestamp = Some(end_timestamp);
+        self
+    }
+
+    pub fn window_size(mut self, window_size: Duration) -> Self {
+        self.window_size = window_size;
+        self
+    }
+
+    pub fn safety_interval(mut self, safety_interval: Duration) -> Self {
+        self.safety_interval = safety_interval;
+        self
+    }
+
+    pub fn sleep_interval(mut self, sleep_interval: Duration) -> Self {
+        self.sleep_interval = sleep_interval;
+        self
+    }
+
+    pub fn readers(mut self, readers: Vec<Arc<StreamReader>>) -> Self {
+        self.readers = readers;
+        self
+    }
+
+    pub fn consumer_factory(mut self, consumer_factory: Arc<dyn ConsumerFactory>) -> Self {
+        self.consumer_factory = Some(consumer_factory);
+        self
+    }
+
+    pub async fn build(self) -> anyhow::Result<(CDCLogReader, RemoteHandle<anyhow::Result<()>>)> {
+        if self.table_name.is_none() {
+            return Err(anyhow!(NewCDCLogReaderError::TableName.as_str()));
```

Comment:
The new error type is completely unnecessary if you are always converting it to string anyway. Either return the new error type in the `Result<(..., ...), NewCDCLogReaderError>`, or get rid of it and inline the strings.